src/avro/writer.rs

```diff
@@ -250,11 +273,18 @@ impl<W: Write> Writer<W> {
 
     /// Create an Avro header based on schema, codec and sync marker.
     fn header(&self) -> Result<Vec<u8>, Error> {
+        self.header_inner(true)
+    }
+
+    /// Like [`header`], but configurable
+    fn header_inner(&self, include_codec: bool) -> Result<Vec<u8>, Error> {
         let schema_bytes = serde_json::to_string(&self.schema)?.into_bytes();
 
         let mut metadata = HashMap::with_capacity(2);
         metadata.insert("avro.schema", Value::Bytes(schema_bytes));
-        metadata.insert("avro.codec", self.codec.avro());
```

Comment:
We should error here if `self.codec` is something other than null, because that is the only case in which omitting the `avro.codec` field is valid.